In [1]:
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

 # Dataset 1

In [2]:
# ri stands for Rhode Island
ri = pd.read_csv('police.csv')
ri.head()

stop_date stop_time  county_name driver_gender  driver_age_raw  \
0  2005-01-02     01:55          NaN             M          1985.0   
1  2005-01-18     08:15          NaN             M          1965.0   
2  2005-01-23     23:15          NaN             M          1972.0   
3  2005-02-20     17:15          NaN             M          1986.0   
4  2005-03-14     10:00          NaN             F          1984.0   

   driver_age driver_race     violation_raw violation  search_conducted  \
0        20.0       White          Speeding  Speeding             False   
1        40.0       White          Speeding  Speeding             False   
2        33.0       White          Speeding  Speeding             False   
3        19.0       White  Call for Service     Other             False   
4        21.0       White          Speeding  Speeding             False   

  search_type   stop_outcome is_arrested stop_duration  drugs_related_stop  
0         NaN       Citation       False      0-15 Min               False  
1         NaN       Citation       False      0-15 Min               False  
2         NaN       Citation       False      0-15 Min               False  
3         NaN  Arrest Driver        True     16-30 Min               False  
4         NaN       Citation       False      0-15 Min               False

In [3]:
ri.shape

(91741, 15)

In [4]:
ri.dtypes

stop_date              object
stop_time              object
county_name           float64
driver_gender          object
driver_age_raw        float64
driver_age            float64
driver_race            object
violation_raw          object
violation              object
search_conducted         bool
search_type            object
stop_outcome           object
is_arrested            object
stop_duration          object
drugs_related_stop       bool
dtype: object

In [5]:
# NaN indicates a missing value? but WHY? 1) not recorded at time, 2) or irevlenat to be filled in
# what is reason why we might fill a cell with a NaN ... avoid messing up date


In [6]:
ri.isnull().head()

stop_date  stop_time  county_name  driver_gender  driver_age_raw  \
0      False      False         True          False           False   
1      False      False         True          False           False   
2      False      False         True          False           False   
3      False      False         True          False           False   
4      False      False         True          False           False   

   driver_age  driver_race  violation_raw  violation  search_conducted  \
0       False        False          False      False             False   
1       False        False          False      False             False   
2       False        False          False      False             False   
3       False        False          False      False             False   
4       False        False          False      False             False   

   search_type  stop_outcome  is_arrested  stop_duration  drugs_related_stop  
0         True         False        False          False               False  
1         True         False        False          False               False  
2         True         False        False          False               False  
3         True         False        False          False               False  
4         True         False        False          False               False

In [7]:
ri.isnull().sum()

stop_date                 0
stop_time                 0
county_name           91741
driver_gender          5335
driver_age_raw         5327
driver_age             5621
driver_race            5333
violation_raw          5333
violation              5333
search_conducted          0
search_type           88545
stop_outcome           5333
is_arrested            5333
stop_duration          5333
drugs_related_stop        0
dtype: int64

In [8]:
True == 1

True

In [9]:
False == 0

True

## 1. Remove column contains missing values

In [10]:
ri.dropna(axis='columns', how='all', inplace = True)

In [11]:
ri.head()

stop_date stop_time driver_gender  driver_age_raw  driver_age driver_race  \
0  2005-01-02     01:55             M          1985.0        20.0       White   
1  2005-01-18     08:15             M          1965.0        40.0       White   
2  2005-01-23     23:15             M          1972.0        33.0       White   
3  2005-02-20     17:15             M          1986.0        19.0       White   
4  2005-03-14     10:00             F          1984.0        21.0       White   

      violation_raw violation  search_conducted search_type   stop_outcome  \
0          Speeding  Speeding             False         NaN       Citation   
1          Speeding  Speeding             False         NaN       Citation   
2          Speeding  Speeding             False         NaN       Citation   
3  Call for Service     Other             False         NaN  Arrest Driver   
4          Speeding  Speeding             False         NaN       Citation   

  is_arrested stop_duration  drugs_related_stop  
0       False      0-15 Min               False  
1       False      0-15 Min               False  
2       False      0-15 Min               False  
3        True     16-30 Min               False  
4       False      0-15 Min               False

### NB** solution2 - providing you noticed the county_name =  91741 , same as the total dataframe shape



In [12]:
# ri.drop('county_name', axis='columns', inplace=True) OR YOU could .. del ri['county_name']

In [13]:
ri.shape #notice one column less

(91741, 14)

In [14]:
ri.columns

Index(['stop_date', 'stop_time', 'driver_gender', 'driver_age_raw',
       'driver_age', 'driver_race', 'violation_raw', 'violation',
       'search_conducted', 'search_type', 'stop_outcome', 'is_arrested',
       'stop_duration', 'drugs_related_stop'],
      dtype='object')

### 2. Do men or Women speed more often?

In [15]:
## driver_gender and violation

In [16]:
speeding_bool = ri['violation'] == 'Speeding'
speeding_bool.head()


0     True
1     True
2     True
3    False
4     True
Name: violation, dtype: bool

In [17]:
speeding_bool.value_counts()

True     48463
False    43278
Name: violation, dtype: int64

In [18]:
speedRI = ri[speeding_bool]['driver_gender'].value_counts()

In [19]:
# speedRI['driver_gender'].value_counts()

In [20]:
speedRI

M    32979
F    15482
Name: driver_gender, dtype: int64

In [21]:
gender_boolM = ri['driver_gender'] == 'M'
gender_boolF = ri['driver_gender'] == 'F'

In [22]:
maleOffence = ri[gender_bool]['violation'].value_counts()
femaleOffence = ri[gender_boolF]['violation'].value_counts()

NameError: name 'gender_bool' is not defined

In [ ]:
maleOffence

In [ ]:
type(femaleOffence)

In [ ]:
femaleOffence

### His answer

In [ ]:
ri[ri.violation == 'Speeding'].driver_gender.value_counts()

In [ ]:
ri[ri.violation == 'Speeding'].driver_gender.value_counts(normalize=True)

### another way to look at it?

In [ ]:
ri[ri.driver_gender == 'M'].violation.value_counts(normalize = True)

In [ ]:
ri[ri.driver_gender == 'M'].violation.value_counts()

In [ ]:
ri[ri.driver_gender == 'F'].violation.value_counts()

### oneliner

In [ ]:
ri.groupby('driver_gender').violation.value_counts()

In [ ]:
ri.groupby('driver_gender').violation.value_counts(normalize=True)

In [ ]:
ri.groupby('driver_gender').violation.value_counts()

In [ ]:
ri.groupby('driver_gender').violation.value_counts(normalize=True).loc[:, 'Speeding']

In [ ]:
ri.groupby('driver_gender').violation.value_counts(normalize=True).unstack()

In [24]:
ri.groupby('driver_gender').agg({'violation' : 'count'})
ri.head()

stop_date stop_time driver_gender  driver_age_raw  driver_age driver_race  \
0  2005-01-02     01:55             M          1985.0        20.0       White   
1  2005-01-18     08:15             M          1965.0        40.0       White   
2  2005-01-23     23:15             M          1972.0        33.0       White   
3  2005-02-20     17:15             M          1986.0        19.0       White   
4  2005-03-14     10:00             F          1984.0        21.0       White   

      violation_raw violation  search_conducted search_type   stop_outcome  \
0          Speeding  Speeding             False         NaN       Citation   
1          Speeding  Speeding             False         NaN       Citation   
2          Speeding  Speeding             False         NaN       Citation   
3  Call for Service     Other             False         NaN  Arrest Driver   
4          Speeding  Speeding             False         NaN       Citation   

  is_arrested stop_duration  drugs_related_stop  
0       False      0-15 Min               False  
1       False      0-15 Min               False  
2       False      0-15 Min               False  
3        True     16-30 Min               False  
4       False      0-15 Min               False

### 3. Does Gender affect who gets searched during a stop?


In [28]:
ri.groupby('driver_gender').search_conducted.value_counts(normalize=True)

driver_gender  search_conducted
F              False               0.979967
               True                0.020033
M              False               0.956674
               True                0.043326
Name: search_conducted, dtype: float64

In [42]:
ri.groupby('driver_gender').search_conducted.value_counts()

driver_gender  search_conducted
F              False               23040
               True                  471
M              False               60170
               True                 2725
Name: search_conducted, dtype: int64

In [41]:
ri[ri.search_conducted == True].driver_gender.value_counts(normalize=True)

M    0.852628
F    0.147372
Name: driver_gender, dtype: float64

### his answer - simplify


In [45]:
#find overall %  likelihood of being searched regardless of Gender  0.34%
ri.search_conducted.value_counts(normalize=True)

False    0.965163
True     0.034837
Name: search_conducted, dtype: float64

In [47]:
ri.search_conducted.mean() #this on a bool col means you get avg mean of 1's and 0's

0.03483720473942948

In [46]:
# % of males vs females
ri.driver_gender.value_counts(normalize=True)

M    0.727901
F    0.272099
Name: driver_gender, dtype: float64

In [49]:
ri.groupby('driver_gender').search_conducted.mean() #one liner!! automatically taking % of the trues

driver_gender
F    0.020033
M    0.043326
Name: search_conducted, dtype: float64

In [50]:
## test more detail re: genders and type of crime

In [51]:
ri.groupby(['violation', 'driver_gender']).search_conducted.mean() #order matters re: violation vs driver_gender

violation            driver_gender
Equipment            F                0.042622
                     M                0.070081
Moving violation     F                0.036205
                     M                0.059831
Other                F                0.056522
                     M                0.047146
Registration/plates  F                0.066140
                     M                0.110376
Seat belt            F                0.012598
                     M                0.037980
Speeding             F                0.008720
                     M                0.024925
Name: search_conducted, dtype: float64

In [52]:
ri.groupby([ 'driver_gender', 'violation']).search_conducted.mean() 

driver_gender  violation          
F              Equipment              0.042622
               Moving violation       0.036205
               Other                  0.056522
               Registration/plates    0.066140
               Seat belt              0.012598
               Speeding               0.008720
M              Equipment              0.070081
               Moving violation       0.059831
               Other                  0.047146
               Registration/plates    0.110376
               Seat belt              0.037980
               Speeding               0.024925
Name: search_conducted, dtype: float64

In [54]:
ri.groupby(['violation', 'driver_gender']).search_conducted.mean().unstack()

driver_gender               F         M
violation                              
Equipment            0.042622  0.070081
Moving violation     0.036205  0.059831
Other                0.056522  0.047146
Registration/plates  0.066140  0.110376
Seat belt            0.012598  0.037980
Speeding             0.008720  0.024925

In [55]:
ri.groupby([ 'driver_gender', 'violation']).search_conducted.mean().unstack()

violation      Equipment  Moving violation     Other  Registration/plates  \
driver_gender                                                               
F               0.042622          0.036205  0.056522             0.066140   
M               0.070081          0.059831  0.047146             0.110376   

violation      Seat belt  Speeding  
driver_gender                       
F               0.012598  0.008720  
M               0.037980  0.024925

In [56]:
a = ri.groupby([ 'driver_gender', 'violation']).search_conducted.mean().unstack()

In [61]:
a = a.round(2)

In [59]:
a

violation      Equipment  Moving violation  Other  Registration/plates  \
driver_gender                                                            
F                   0.04              0.04   0.06                 0.07   
M                   0.07              0.06   0.05                 0.11   

violation      Seat belt  Speeding  
driver_gender                       
F                   0.01      0.01  
M                   0.04      0.02

### In summary:  causation is difficult to conclude  so focus on relationships

#### include all relevant factors when studying a relationship!

## 4. Why is search_type missing so often?1 

In [62]:
ri.isnull().sum()  #this is number of missing values in each column!

stop_date                 0
stop_time                 0
driver_gender          5335
driver_age_raw         5327
driver_age             5621
driver_race            5333
violation_raw          5333
violation              5333
search_conducted          0
search_type           88545
stop_outcome           5333
is_arrested            5333
stop_duration          5333
drugs_related_stop        0
dtype: int64

In [63]:
ri.search_conducted.value_counts() #evidence that reason 88k above "search_type" is missing above! ie not actually searched

False    88545
True      3196
Name: search_conducted, dtype: int64

In [69]:
ri.search_type.value_counts(dropna=False) #NB not a T F value - 88k were NOT searched depsite having BEEN stopped! 

NaN                                                         88545
Incident to Arrest                                           1219
Probable Cause                                                891
Inventory                                                     220
Reasonable Suspicion                                          197
Protective Frisk                                              161
Incident to Arrest,Inventory                                  129
Incident to Arrest,Probable Cause                             106
Probable Cause,Reasonable Suspicion                            75
Incident to Arrest,Inventory,Probable Cause                    34
Probable Cause,Protective Frisk                                33
Incident to Arrest,Protective Frisk                            33
Inventory,Probable Cause                                       22
Incident to Arrest,Reasonable Suspicion                        13
Protective Frisk,Reasonable Suspicion                          11
Incident t

In [68]:
ri[ri.search_conducted ==False].search_type.value_counts(dropna=False) #shows that by default ie true, all NaN/null are dropped ie value NOT 0

NaN    88545
Name: search_type, dtype: int64

In [70]:
##try to always verify assumptions!! :P - think about your NULLS/

## During a search how often is driver frisked?
Search_type.value_counts has "protective search" ;)

In [79]:
ri.dtypes['search_type']

dtype('O')

In [95]:
ri[ri.search_conducted == True].search_type.value_counts()

Incident to Arrest                                          1219
Probable Cause                                               891
Inventory                                                    220
Reasonable Suspicion                                         197
Protective Frisk                                             161
Incident to Arrest,Inventory                                 129
Incident to Arrest,Probable Cause                            106
Probable Cause,Reasonable Suspicion                           75
Incident to Arrest,Inventory,Probable Cause                   34
Incident to Arrest,Protective Frisk                           33
Probable Cause,Protective Frisk                               33
Inventory,Probable Cause                                      22
Incident to Arrest,Reasonable Suspicion                       13
Inventory,Protective Frisk                                    11
Incident to Arrest,Inventory,Protective Frisk                 11
Protective Frisk,Reasonab

In [121]:
ri[ri.search_conducted == True]

stop_date stop_time driver_gender  driver_age_raw  driver_age  \
24     2005-08-28     01:00             M          1979.0        26.0   
40     2005-10-01     00:00             M          1988.0        17.0   
41     2005-10-01     00:00             M          1988.0        17.0   
80     2005-10-02     09:30             M          1975.0        30.0   
106    2005-10-03     14:00             M          1978.0        27.0   
118    2005-10-04     01:20             M          1956.0        49.0   
139    2005-10-05     00:20             M          1980.0        25.0   
163    2005-10-05     22:50             M          1980.0        25.0   
208    2005-10-07     12:00             M          1947.0        58.0   
209    2005-10-07     12:45             F          1981.0        24.0   
210    2005-10-07     13:00             M          1987.0        18.0   
219    2005-10-07     23:30             M          1986.0        19.0   
246    2005-10-08     20:40             M          1987.0        18.0   
293    2005-10-10     16:00             M          1955.0        50.0   
308    2005-10-11     03:15             M          1964.0        41.0   
315    2005-10-11     13:40             M          1963.0        42.0   
323    2005-10-11     23:49             M          1953.0        52.0   
337    2005-10-12     20:20             M          1977.0        28.0   
339    2005-10-12     20:30             M          1987.0        18.0   
340    2005-10-12     20:30             M          1987.0        18.0   
377    2005-10-16     00:00             M          1981.0        24.0   
381    2005-10-16     02:50             M          1976.0        29.0   
388    2005-10-16     10:40             M          1961.0        44.0   
401    2005-10-16     19:30             M          1977.0        28.0   
415    2005-10-17     09:30             M          1983.0        22.0   
435    2005-10-18     02:45             M          1967.0        38.0   
462    2005-10-18     23:46             M          1958.0        47.0   
465    2005-10-19     01:00             M          1956.0        49.0   
488    2005-10-20     01:52             F          1986.0        19.0   
512    2005-10-21     06:05             M          1961.0        44.0   
...           ...       ...           ...             ...         ...   
90382  2015-10-30     08:38             M          1961.0        54.0   
90449  2015-10-31     16:05             M          1989.0        26.0   
90530  2015-11-02     22:55             M          1980.0        35.0   
90539  2015-11-03     02:04             M          1981.0        34.0   
90540  2015-11-03     02:04             M          1981.0        34.0   
90548  2015-11-03     13:13             F          1991.0        24.0   
90567  2015-11-04     10:11             M          1983.0        32.0   
90726  2015-11-11     13:52             M          1978.0        37.0   
90782  2015-11-14     00:10             M          1983.0        32.0   
90810  2015-11-15     13:23             M          1995.0        20.0   
90817  2015-11-16     00:22             F          1986.0        29.0   
90864  2015-11-18     15:34             M          1990.0        25.0   
90938  2015-11-22     00:36             M          1983.0        32.0   
91027  2015-11-25     21:53             M          1983.0        32.0   
91034  2015-11-26     02:47             M          1980.0        35.0   
91101  2015-11-29     10:29             M          1991.0        24.0   
91146  2015-12-01     06:47             M          1960.0        55.0   
91155  2015-12-01     18:07             M          1986.0        29.0   
91165  2015-12-02     09:35             M          1983.0        32.0   
91194  2015-12-03     14:48             M          1991.0        24.0   
91253  2015-12-06     10:33             M          1993.0        22.0   
91284  2015-12-07     14:10             M          1976.0        39.0   
91384  2015-12-12     16:46             M          1978.0        3

In [113]:
ri.search_type.str.contains("Protective Frisk").value_counts()

False    2922
True      274
Name: search_type, dtype: int64

In [122]:
274 + 2922

3196

In [135]:
3196/100  ## could be expressed as 274 / 3196 rather than unessecary next step 

31.96

In [134]:
274/31.96

8.573216520650814

### our answer above is hacky! :P

#better way of doing it! - pandas has string methods apply across series BUT python also has string methods

In [126]:
ri.search_type.str.contains("Protective Frisk").head()

0    NaN
1    NaN
2    NaN
3    NaN
4    NaN
Name: search_type, dtype: object

In [127]:
ri['frisk'] = ri.search_type.str.contains("Protective Frisk")

In [129]:
ri['frisk'].value_counts(dropna=False)

NaN      88545
False     2922
True       274
Name: frisk, dtype: int64

In [130]:
ri.frisk.sum()

274

In [132]:
ri.frisk.mean()  #I forgot THIS! returns a % of TRUE values (true =1 , false = 0)

0.08573216520650813

In [133]:
# or you could also write
274 / (274 + 2922)

0.08573216520650813